internal/driver/mobile/canvas.go

```diff
@@ -10,124 +10,132 @@ import (
 	"fyne.io/fyne/v2/container"
 	"fyne.io/fyne/v2/driver/mobile"
 	"fyne.io/fyne/v2/internal/app"
-	"fyne.io/fyne/v2/internal/driver"
+	intdriver "fyne.io/fyne/v2/internal/driver"
 	"fyne.io/fyne/v2/internal/driver/common"
 	"fyne.io/fyne/v2/theme"
 	"fyne.io/fyne/v2/widget"
 )
 
-var _ fyne.Canvas = (*mobileCanvas)(nil)
+var _ fyne.Canvas = (*canvas)(nil)
 
-type mobileCanvas struct {
+type canvas struct {
 	common.Canvas
-
-	content          fyne.CanvasObject
-	windowHead, menu fyne.CanvasObject
-	scale            float32
-	size             fyne.Size
-
-	touched map[int]mobile.Touchable
-	padded  bool
-
-	onTypedRune func(rune)
-	onTypedKey  func(event *fyne.KeyEvent)
-
-	inited                bool
-	lastTapDown           map[int]time.Time
-	lastTapDownPos        map[int]fyne.Position
-	dragging              fyne.Draggable
-	dragStart, dragOffset fyne.Position
-
-	touchTapCount   int
+	content         fyne.CanvasObject
+	device          *device
+	dragOffset      fyne.Position
+	dragStart       fyne.Position
+	dragging        fyne.Draggable
+	initialized     bool
+	lastTapDown     map[int]time.Time
+	lastTapDownPos  map[int]fyne.Position
+	menu            fyne.CanvasObject
+	onTypedKey      func(event *fyne.KeyEvent)
+	onTypedRune     func(rune)
+	padded          bool
+	scale           float32
+	size            fyne.Size
 	touchCancelFunc context.CancelFunc
 	touchLastTapped fyne.CanvasObject
+	touchTapCount   int
+	touched         map[int]mobile.Touchable
+	windowHead      fyne.CanvasObject
 }
 
-// NewCanvas creates a new gomobile mobileCanvas. This is a mobileCanvas that will render on a mobile device using OpenGL.
-func NewCanvas() fyne.Canvas {
-	ret := &mobileCanvas{padded: true}
-	ret.scale = fyne.CurrentDevice().SystemScaleForWindow(nil) // we don't need a window parameter on mobile
-	ret.touched = make(map[int]mobile.Touchable)
-	ret.lastTapDownPos = make(map[int]fyne.Position)
-	ret.lastTapDown = make(map[int]time.Time)
+func newCanvas(dev fyne.Device) fyne.Canvas {
+	d, _ := dev.(*device)
+	ret := &canvas{
+		Canvas: common.Canvas{
+			OnFocus:   handleKeyboard,
+			OnUnfocus: hideVirtualKeyboard,
+		},
+		device:         d,
```

Comment:
No need to do the interface cast outside and ignore the `ok` parameter. You can just do an explicit cast inline:
```suggestion
	ret := &canvas{
		Canvas: common.Canvas{
			OnFocus:   handleKeyboard,
			OnUnfocus: hideVirtualKeyboard,
		},
		device:         dev.(*device),
```

Reply:
I’m quite sure that your proposal will crash if `dev` would be a test device.

Reply:
`x, _ := a.(*typ)` and `x := a.(*typ)` are semantically different.

Reply:
Surely your code will crash as well given that you don’t check the “ok” value (second parameter)?

Reply:
Never mind. Of course you get a nil in that case and any other codes does a nil check before checking. Right...